### PR TITLE
[#1150] Part 2: User Profile Enhancements for Former Names 

### PR DESCRIPTION
**Pull request by jhflorey**
This PR addresses issue https://github.com/episphere/connect/issues/1150

### CODE REVIEW

File: i18n/en.js
Comment (amber-emmes) [inline i18n/en.js:102]: Should `formerNameIntrodution` be corrected to `formerNameIntroduction`?

File: i18n/es.js
Comment (amber-emmes) [inline i18n/es.js:103]: Same comment.

File: js/components/form.js
Comment (amber-emmes) [inline js/components/form.js:69]: Same comment.

**Review by amber-emmes — COMMENTED** [review]
Comments added.

File: i18n/en.js
Comment (jhflorey) [reply i18n/en.js]: Resolved

File: i18n/es.js
Comment (jhflorey) [reply i18n/es.js]: Resolved

File: js/components/form.js
Comment (jhflorey) [reply js/components/form.js]: Resolved

File: js/components/form.js
Comment (JoeArmani) [inline js/components/form.js:6]: state of birth can use `allStates` (shared.js) and be a dropdown selector to ensure no typos
country of birth can use `allCountries` (shared.js) and be a dropdown selector for the same reason.

This makes it more accurate when researchers quantify results over time.

There are examples of the allStates implementation in the existing code, and the allCountries implementation should basically be the same thing.

File: js/settingsHelpers.js
Comment (JoeArmani) [inline js/settingsHelpers.js:1009]: I really think we need to match these data structures for long-term compatibility.

Here's a screenshot of the data strucure this generated (from Firestore dev):

Screenshot 2024-11-20 at 4 10 54 PM

---
Here's an example of the existing data structure for the user profile history spec:
Screenshot 2024-11-20 at 4 12 18 PM

**Review by JoeArmani — CHANGES_REQUESTED** [review]
Left some comments.

•The state and country dropdowns (from a previous PR) are helpful for long-term analytics

•I think it's important to match the existing UP History data structure so it can be:
(1) Implemented in the user profile in a consistent way
(2) Reliably accessed when needed

File: js/components/form.js
Comment (jhflorey) [reply js/components/form.js:6]: This is confirmed by Rohan

image

File: js/settingsHelpers.js
Comment (jhflorey) [reply js/settingsHelpers.js:1009]: Here is the result after using the code change.
image

File: js/components/form.js
Comment (JoeArmani) [reply js/components/form.js:6]: sounds good, thanks

File: js/settingsHelpers.js
Comment (JoeArmani) [reply js/settingsHelpers.js:1009]: I see mixed data issues in the `[Array]` parts. Those are strings in the existing UP history implementation.

We are starting to use that UP history for some things like NORC Birthday cards, and we need to make sure the data is in a predictable format.

There will be more entries in the array, but I think those should be expanded out so each entry for 231676651, 399159511, and 996038075 is a separate object in the 569151507 UP History array.

Each UP history object needs:
•update string (first/middle/last name).
•update timestamp 371303487.
•profile change requested by 611005658 (this will be the participant's email address for this case).

Comment (jhflorey) [reply js/settingsHelpers.js]: Just updated, please check it again.

File: app.js
Comment (JoeArmani) [inline app.js:21]: Please disable this before merging

**Review by JoeArmani — CHANGES_REQUESTED** [review]
Confirming the UP history data structure now matches the existing spec. Thank you.

I ran into an issue in testing where I:
(1) Filled out the form
(2) Clicked the submit button
(3) Viewed the verification form
(4) Clicked that submit button

Everything worked well to that point.

Then, the data saved correctly, but the form did not advance to the next screen.

Please test this and I'll approve once you're sure that part is working.

File: app.js
Comment (jhflorey) [reply app.js]: Solved

**Review by JoeArmani — APPROVED** [review]

**Pull request merged**
